### ISSUE TITLE
design: new event or state to indicate when non-compute job tasks are complete

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#5774]
As noted in https://github.com/flux-framework/flux-coral2/issues/137, there's an issue when completion of a job depends on work other than the compute tasks being managed by the job shell.

Here's a brief summary of the issues as I understand them:

 - The `finish` event transitions a job from RUN to CLEANUP (See [RFC 21](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_21.html)), but the `finish` event only indicates that job shells have completed, not that any other job related work has also completed (e.g. in this case, data movement)
 - Tools like `flux job attach` and `flux watch` consider a job finished at the `finish` event to avoid waiting for cleanup tasks like the job epilog, which may take several minutes or could hang indefinitely in some cases.
 - Work such as data movement, post-processing, etc requires a way to be initiated after the compute or main portion of the job work is done. Currently, the rabbit solution uses the transition to CLEANUP to intiate this work via an `epilog-start` event. But the RFC 21 definition of CLEANUP is "The job has completed or an exception has occurred." Ideally work that prevents a job from fully completing should not occur in a state that says the job has completed.
 - Work as described in this use case may or may not tie up compute resources assigned to the job. It would be nice if that work could run in parallel with both the job itself and any other administrative epilog work, and could denote which, if any resources should be held back from the scheduler until the work in complete. (In fact, the work may not complete until after all resources are freed, e.g. if data movement or post processing is slow and node epilogs are fast)
- When run as an epilog, data movement tasks do not have a way to influence the job "result" (pass/fail/etc.) Perhaps this isn't desired in all cases, but it would be nice to have a way to have a job marked as FAILED if the data movement failed.

I'll just throw out one idea to get the discussion started:
- Support a new `complete` event that is emitted by the job manager once all completion tasks are emitted.
- The `finish` event is a special case of a  completion task, it becomes required for completion by the `start` event
- Other completion tasks are registered by emitting a special event, maybe something like `completion-requires event=name ranks=idset`, which adds event `name` to the set of events required for the `complete` event, with optional `ranks` key including an idset of execution targets that should be held back from the scheduler until the indicated event. The `start` event would be equivalent internally to `completion-requires event=finish ranks=all`
- Completion events may optionally include a `status=N` key that indicates the status of the completion task. If present, a failure noted in the event should be propagated to the `complete` event `status` value. (Alternative would be to have completion tasks emit a job exception on failure)
- If the set of completion event is empty, then an `exception` event is sufficient for the `complete` event to be emitted.
- The `complete` event may occur after all other events, including epilogs etc, so emission of the `clean` event should be delayed in until this event is posted

There's some obvious drawbacks to the above proposal:
- addition of a new event in even the common case
- the job manager currently tracks a single "end_event" and this muddies those waters
- jobs can be in CLEANUP while some user work is still occurring on resources. It might be nice if there was a way to differentiate administrative cleanup vs other job-related work (I'd hate to propose a new job state though)

**Comment by grondo** [comment]
This issue was probably premature to bring up and is unlikely to get further discussion.

The current proposal is to revert the tools to wait for the `clean` event and eventually divorce the `clean` event from the administrative epilog which can take a long time.
